File: packages/ocp-crypto/src/proofs/chaumPedersen.ts
import type { GroupElement } from "../utils/group.js";
import {
  groupElementFromBytes,
  groupElementToBytes,
  mulBase,
  mulPoint,
  pointAdd,
  pointEq,
} from "../utils/group.js";
import type { Scalar } from "../utils/scalar.js";
import {
  assertScalar,
  scalarAdd,
  scalarFromBytes,
  scalarMul,
  scalarToBytes,
} from "../utils/scalar.js";
import { Transcript } from "../utils/transcript.js";
import { concatBytes } from "../utils/bytes.js";

export type ChaumPedersenProof = {
  // a = w*G
  a: GroupElement;
  // b = w*c1
  b: GroupElement;
  // s = w + e*x
  s: Scalar;
};

const CP_DOMAIN = "ocp/v1/chaum-pedersen-eqdl";

export function chaumPedersenProve(params: {
  // Public statement:
  y: GroupElement; // y = x*G
  c1: GroupElement;
  d: GroupElement; // d = x*c1
  // Witness:
  x: Scalar;
  // Nonce:
  w: Scalar;
}): ChaumPedersenProof {
  const { y, c1, d, x, w } = params;
  assertScalar(x);
  assertScalar(w);

  const a = mulBase(w);
  const b = mulPoint(c1, w);

  const tr = new Transcript(CP_DOMAIN);
  tr.appendMessage("y", groupElementToBytes(y));
  tr.appendMessage("c1", groupElementToBytes(c1));
  tr.appendMessage("d", groupElementToBytes(d));
  tr.appendMessage("a", groupElementToBytes(a));
  tr.appendMessage("b", groupElementToBytes(b));
  const e = tr.challengeScalar("e");

  const s = scalarAdd(w, scalarMul(e, x));
  return { a, b, s };
}

export function chaumPedersenVerify(params: {
  y: GroupElement;
  c1: GroupElement;
  d: GroupElement;
  proof: ChaumPedersenProof;
}): boolean {
  const { y, c1, d, proof } = params;
  const { a, b, s } = proof;

  assertScalar(s);

  const tr = new Transcript(CP_DOMAIN);
  tr.appendMessage("y", groupElementToBytes(y));
  tr.appendMessage("c1", groupElementToBytes(c1));
  tr.appendMessage("d", groupElementToBytes(d));
  tr.appendMessage("a", groupElementToBytes(a));
  tr.appendMessage("b", groupElementToBytes(b));
  const e = tr.challengeScalar("e");

  // Check: s*G == a + e*y
  const lhs1 = mulBase(s);
  const rhs1 = pointAdd(a, mulPoint(y, e));
  if (!pointEq(lhs1, rhs1)) return false;

  // Check: s*c1 == b + e*d
  const lhs2 = mulPoint(c1, s);
  const rhs2 = pointAdd(b, mulPoint(d, e));
  if (!pointEq(lhs2, rhs2)) return false;

  return true;
}

// Encoding: A(32) || B(32) || s(32 le)
export function encodeChaumPedersenProof(p: ChaumPedersenProof): Uint8Array {
  return concatBytes(groupElementToBytes(p.a), groupElementToBytes(p.b), scalarToBytes(p.s));
}

export function decodeChaumPedersenProof(bytes: Uint8Array): ChaumPedersenProof {
  if (!(bytes instanceof Uint8Array) || bytes.length !== 32 + 32 + 32) {
    throw new Error("decodeChaumPedersenProof: expected 96 bytes");
  }
  const aBytes = bytes.slice(0, 32);
  const bBytes = bytes.slice(32, 64);
  const sBytes = bytes.slice(64, 96);

  const a = groupElementFromBytes(aBytes);
  const b = groupElementFromBytes(bBytes);
  // scalarFromBytes expects 32-bytes canonical; but `s` is always mod q so it should be canonical.
  // Keep decode strict to match acceptance tests.
  const s = scalarFromBytes(sBytes);
  return { a, b, s };
}

File: apps/chain/internal/ocpcrypto/chaum_pedersen.go
package ocpcrypto

import "fmt"

type ChaumPedersenProof struct {
	// a = w*G
	A Point
	// b = w*c1
	B Point
	// s = w + e*x
	S Scalar
}

const chaumPedersenDomain = "ocp/v1/chaum-pedersen-eqdl"

func ChaumPedersenProve(y Point, c1 Point, d Point, x Scalar, w Scalar) (ChaumPedersenProof, error) {
	if x.IsZero() {
		// witness x=0 is allowed; keep this for parity with TS which only checks scalar range.
	}
	if w.IsZero() {
		return ChaumPedersenProof{}, fmt.Errorf("chaum-pedersen: w must be non-zero")
	}

	a := MulBase(w)
	b := MulPoint(c1, w)

	tr := NewTranscript(chaumPedersenDomain)
	_ = tr.AppendMessage("y", y.Bytes())
	_ = tr.AppendMessage("c1", c1.Bytes())
	_ = tr.AppendMessage("d", d.Bytes())
	_ = tr.AppendMessage("a", a.Bytes())
	_ = tr.AppendMessage("b", b.Bytes())
	e, err := tr.ChallengeScalar("e")
	if err != nil {
		return ChaumPedersenProof{}, err
	}

	s := ScalarAdd(w, ScalarMul(e, x))
	return ChaumPedersenProof{A: a, B: b, S: s}, nil
}

func ChaumPedersenVerify(y Point, c1 Point, d Point, proof ChaumPedersenProof) (bool, error) {
	tr := NewTranscript(chaumPedersenDomain)
	_ = tr.AppendMessage("y", y.Bytes())
	_ = tr.AppendMessage("c1", c1.Bytes())
	_ = tr.AppendMessage("d", d.Bytes())
	_ = tr.AppendMessage("a", proof.A.Bytes())
	_ = tr.AppendMessage("b", proof.B.Bytes())
	e, err := tr.ChallengeScalar("e")
	if err != nil {
		return false, err
	}

	// Check: s*G == a + e*y
	lhs1 := MulBase(proof.S)
	rhs1 := PointAdd(proof.A, MulPoint(y, e))
	if !PointEq(lhs1, rhs1) {
		return false, nil
	}

	// Check: s*c1 == b + e*d
	lhs2 := MulPoint(c1, proof.S)
	rhs2 := PointAdd(proof.B, MulPoint(d, e))
	if !PointEq(lhs2, rhs2) {
		return false, nil
	}
	return true, nil
}

// Encoding: A(32) || B(32) || s(32 le)
func EncodeChaumPedersenProof(p ChaumPedersenProof) []byte {
	return concatBytes(p.A.Bytes(), p.B.Bytes(), p.S.Bytes())
}

func DecodeChaumPedersenProof(b []byte) (ChaumPedersenProof, error) {
	if len(b) != 96 {
		return ChaumPedersenProof{}, fmt.Errorf("chaum-pedersen: expected 96 bytes")
	}
	a, err := PointFromBytesCanonical(b[0:32])
	if err != nil {
		return ChaumPedersenProof{}, err
	}
	bl, err := PointFromBytesCanonical(b[32:64])
	if err != nil {
		return ChaumPedersenProof{}, err
	}
	s, err := ScalarFromBytesCanonical(b[64:96])
	if err != nil {
		return ChaumPedersenProof{}, err
	}
	return ChaumPedersenProof{A: a, B: bl, S: s}, nil
}

